cpp/CMakeLists.txt

```diff
@@ -35,16 +37,34 @@ endif(CMAKE_COMPILER_IS_GNUCXX)
 set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -D_GLIBCXX_USE_CXX11_ABI=0")
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -D_GLIBCXX_USE_CXX11_ABI=0")
 
-#set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -gencode=arch=compute_60,code=sm_60 -gencode=arch=compute_61,code=sm_61")
-set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -gencode=arch=compute_60,code=sm_60")
-set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_70,code=compute_70")
+# e.g. cmake -DGPU_COMPUTE_CAPABILITY=60 ... to support GPU Compute Capability 6.0
+if(NOT GPU_COMPUTE_CAPABILITY)
+    message(AUTHOR_WARNING "NVIDIA GPU Compute Capability is not defined, using defult NVIDIA GPU compute capability version 7.0")
+    set(GPU_COMPUTE_CAPABILITY "70")
+endif ()
+
+# GPU Compute Capability: virtual architecture (code=compute_X)
+set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -gencode arch=compute_${GPU_COMPUTE_CAPABILITY},code=compute_${GPU_COMPUTE_CAPABILITY}")
+# GPU Compute Capability: real architecture (code=sm_X)
+set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -gencode arch=compute_${GPU_COMPUTE_CAPABILITY},code=sm_${GPU_COMPUTE_CAPABILITY}")
 
 set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} --expt-extended-lambda --expt-relaxed-constexpr")
 
 # suppress SHFL warnings caused by Modern GPU
 # TODO: remove this when Modern GPU is removed or fixed to use shfl_sync
 set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -Werror cross-execution-space-call -Wno-deprecated-declarations -Xptxas --disable-warnings")
 
+# e.g. cmake -DGPU_PTXAS_MAXRREGCOUNT=48 ... to use nvcc ptxas flags: -Xptxas --maxrregcount=48
+if(GPU_PTXAS_MAXRREGCOUNT)
+    message(STATUS "NVIDIA ptxas flags for nvcc: -Xptxas --maxrregcount=${GPU_PTXAS_MAXRREGCOUNT}")
+    set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -Xptxas --maxrregcount=${GPU_PTXAS_MAXRREGCOUNT}")
+else()
+    # unoptimised ptx code
+    set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} -Xptxas")
+endif ()
+
+set(CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS} --disable-warnings")
```

Comment:
I'm against disabling warnings (except when there's an artificial and ubiquitous one, which clutters the terminal);. Note that, on the host, not only do we keep warnings, we are even stricter and make errors out of the warnings.

Reply:
Not just on the host. My philosophy is "Warnings ARE errors", which is why I've enabled warnings as errors everywhere possible (for some reason nvcc doesn't always treat all warnings as errors even when you enable that setting).

Reply:
Please eliminate this change.

Reply:
In this case, you should disable the specific warning using a pragma that is only active in the smallest possible scope you need to solve the issue.  The trickiest case is when the warning is caused by including a file from a package you don't control. This has happened to me with nvcc and the Arrow generated IPC headers, which I solved like this: https://github.com/rapidsai/cudf/blob/c9fbc8b275315d173e4345e10283b6c697fa7556/cpp/src/comms/ipc/ipc.cu#L3 (tricky to find documentation of this since the pragma is provided by the EDG compiler front-end used by NVCC).